lib/internal/main/watch_mode.js

```diff
@@ -0,0 +1,132 @@
+const { getOptionValue } = require('internal/options');
+const { emitExperimentalWarning } = require('internal/util');
+const { FilesWatcher } = require('internal/watch_mode/files_watcher');
+const { green, blue, red, white, clear } = require('internal/util/colors');
+
+const { spawn } = require('child_process');
+const { inspect } = require('util');
+const { setTimeout, clearTimeout } = require('timers');
+const { resolve } = require('path');
+const { once, on } = require('events');
+
+
+prepareMainThreadExecution(false, false);
+markBootstrapComplete();
+
+// TODO(MoLow): Make kill signal configurable
+const kKillSignal = 'SIGTERM';
+const kShouldFilterModules = getOptionValue('--watch-path').length === 0;
+const kWatchedPaths = ArrayPrototypeMap(getOptionValue('--watch-path'), (path) => resolve(path));
+const kCommand = ArrayPrototypeSlice(process.argv, 1);
+const kCommandStr = inspect(ArrayPrototypeJoin(kCommand, ' '));
+const args = ArrayPrototypeFilter(process.execArgv, (arg, i, arr) =>
+  arg !== '--watch-path' && arr[i - 1] !== '--watch-path' && arg !== '--watch');
+ArrayPrototypePushApply(args, kCommand);
+
+const watcher = new FilesWatcher({ throttle: 500, mode: kShouldFilterModules ? 'filter' : 'all' });
+ArrayPrototypeForEach(kWatchedPaths, (p) => watcher.watchPath(p));
+
+let graceTimer;
+let child;
+let exited;
+
+function start() {
+  exited = false;
+  const stdio = kShouldFilterModules ? ['inherit', 'inherit', 'inherit', 'ipc'] : undefined;
+  child = spawn(process.execPath, args, { stdio, env: { ...process.env, WATCH_REPORT_DEPENDENCIES: '1' } });
+  watcher.watchChildProcessModules(child);
+  child.once('exit', (code) => {
+    exited = true;
+    if (code === 0) {
+      process.stdout.write(`${blue}Completed running ${kCommandStr}${white}\n`);
+    } else {
+      process.stdout.write(`${red}Failed running ${kCommandStr}${white}\n`);
+    }
+  });
+}
+
+async function killAndWait(signal = kKillSignal, force = false) {
+  child?.removeAllListeners();
+  if (!child) {
+    return;
+  }
+  if ((child.killed || exited) && !force) {
+    return;
+  }
+  const onExit = once(child, 'exit');
+  child.kill(signal);
+  const { 0: exitCode } = await onExit;
+  return exitCode;
+}
+
+function reportGracefulTermination() {
+  // Log if process takes more than 500ms to stop.
+  let reported = false;
+  clearTimeout(graceTimer);
+  graceTimer = setTimeout(() => {
+    reported = true;
+    process.stdout.write(`${blue}Waiting for graceful termination...${white}\n`);
+  }, 500).unref();
+  return () => {
+    clearTimeout(graceTimer);
+    if (reported) {
+      process.stdout.write(`${clear}${green}Gracefully restarted ${kCommandStr}${white}\n`);
+    }
+  };
+}
+
+async function stop() {
+  watcher.clearFileFilters();
+  const clearGraceReport = reportGracefulTermination();
+  await killAndWait();
+  clearGraceReport();
+}
+
+async function restart() {
+  process.stdout.write(`${clear}${green}Restarting ${kCommandStr}${white}\n`);
```

Comment:
This "clear" should be at least configurable, as the `--preserveWatchOutput` option with `tsc`